### ISSUE TITLE
`RuleTester` should offer an API that just runs a single scenario

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#11609]
**Edit:** Remove an irrelevant section.

**The version of ESLint you are using.**

v5.16.0

**The problem you want to solve.**

I'd like to be able to give more informative test names for my individual scenarios, and I'd like to group scenarios at a finer grained level than just the rule level + valid/invalid.

The existing `ruleTester.run(name, rule, {valid, invalid})` API with its assertions requiring both `valid` and `invalid` to be non-empty and invoking `describe(ruleName, () => ...)` instead of letting me controlling it, is getting in the way. For simple rules, this works, but for complex, multi-faceted rules, it makes it much more difficult than it should be.

I'd like to split mine into multiple rules and manage everything through a configurable config preset, but there's a few glitches:

- My "features" don't map cleanly to "rules".
- Plugins don't offer the ability to generate configs programmatically, so I'd end up having to generate *every* possible config combination, something that obviously would blow up *quick*. (For me, it's currently 2^26 * 2 * 12 = 1,610,612,736 possible configs. Yes, over a billion, and that's obviously impractical to implement.)

**Your take on the correct solution to problem.**

Offer a `ruleTester.runSingle(name, rule, {...scenario, isValid})`, so I can use them in individual test cases like `it("does something right", () => ruleTester.run("my-rule", myRule, {code: "...", isValid: true}))`. It also gives me better flexibility on how my tests are structured, so I can test specific features of my tests.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by isiahmeadows** [comment]
I'll note that I currently have a [working workaround](https://github.com/isiahmeadows/eslint-plugin-no-esnext/blob/master/test-util.js) in place for my immediate use case, but with my proposal above, I could simplify it to just this one helper:

```js
"use strict"

const rule = require("./index").rules["no-esnext"]
const {RuleTester} = require("eslint")

const ruleTester = new RuleTester()

exports.check = ({
    code, isModule, errors,
    ecmaVersion, excludeFeatures, includeFeatures,
}) => {
    const options = {ecmaVersion: ecmaVersion || 2019}
    const result = {
        code, options: [options],
        // Use the most recent ECMAScript version available.
        parserOptions: {ecmaVersion: 2019},
        isValid: errors.length === 0,
    }

    if (isModule) result.parserOptions.sourceType = "module"
    if (errors.length !== 0) result.errors = errors
    if (excludeFeatures) options.excludeFeatures = excludeFeatures
    if (includeFeatures) options.includeFeatures = includeFeatures

    ruleTester.runSingle("no-eslint", rule, result)
}

```

**Comment by not-an-aardvark** [comment]
Have you tried overriding `RuleTester.describe` and `RuleTester.it` to something else? It seems like this would allow you to customize the behavior in the way you're asking (for example, you could set `RuleTester.describe` to a no-op, and change `RuleTester.it` to a function that invokes a specific group of test cases).

**Comment by not-an-aardvark** [comment]
Never mind, I just saw your workaround and it seems like that is roughly what you're doing.

But I'm confused about why something like this wouldn't work as a replacement for `runSingle`:

```js
RuleTester.describe = () => {};
RuleTester.it = global.it;
```

It seems like there is a lot of possible customization that people could want to do, and I think we should make it possible for them to do that customization, but it's not obvious to me that something like `runSingle` would be generally useful if there is a simple alternative to accomplish the same thing.

**Comment by isiahmeadows** [comment]
@not-an-aardvark 

> Never mind, I just saw your workaround and it seems like that is roughly what you're doing.
> 
> But I'm confused about why something like this wouldn't work as a replacement for `runSingle`:
> 
> ```js
> RuleTester.describe = () => {};
> RuleTester.it = global.it;
> ```

That won't work when [`ruleTester.run(...)` requires both the `valid` and `invalid` arrays to be non-empty](https://github.com/eslint/eslint/blob/master/lib/testers/rule-tester.js#L283-L287). It obviously makes sense that you want rules to have both valid and invalid scenarios, but it did get in my way of correctly testing by rule. (In particular, [the one "valid" test I have for one of my features](https://github.com/isiahmeadows/eslint-plugin-no-esnext/blob/master/test/array-destructuring.js#L38-L42) is a very low quality test that's just there to satisfy the runner.)

My workaround works by taking the various tasks and enqueueing them to later be passed in the `valid` and `invalid` parameters. I had to also take into account the test names wouldn't be in the same order as I initially receive them, so that added some complexity as well. The state machine was so I could coordinate the titles correctly and so I could ignore the top-level `RuleTester.describe(ruleName, () => ...)`. I found it the easiest and simplest to just rely on call order for the state machine (horribly brittle hack, I know):

```js
// 1.
RuleTester.describe(ruleName, () => {
    // 2.
    RuleTester.describe("valid", () => {
        // ...
    })

    // 3.
    RuleTester.describe("invalid", () => {
        // ...
    })
})
```

> It seems like there is a lot of possible customization that people could want to do, and I think we should make it possible for them to do that customization, but it's not obvious to me that something like runSingle would be generally useful if there is a simple alternative to accomplish the same thing.

The customization side is just me explaining how my rule entails multiple conceptual features, but why I can't readily decompose it to multiple rules to where the standard testing API with its implicit `describe(ruleName, () => ...)` actually works. (If rules could be aware of other rules, that would solve my problem almost entirely and I would be able to fix my code to not need this bug.)

**Comment by not-an-aardvark** [comment]
If I'm not mistaken, `ruleTester.run` requires both the `valid` and `invalid` arrays to exist, but it doesn't require them to be non-empty. If you just want to run a single `invalid` test you should just be able to make the `valid` array empty.

**Comment by isiahmeadows** [comment]
Okay...that's weird because I experienced errors while trying just that. (May have linked to the wrong section, but I experienced that prior to filing this bug.)

**Comment by ilyavolodin** [comment]
I'm not against expending ruleTester to cover those scenarios, but the question that I have is why do you want to use ruleTester for this to begin with? ruleTester was designed to be a wrapper around mocha to cut down on the repetitions of the same code. It was meant to be simple shortcut to verifying standard rule scenarios. It sounds like what you are trying to do is a lot more complicated. So why not just drop ruleTester all together and just create straight-out mocha tests?

**Comment by isiahmeadows** [comment]
@ilyavolodin I could (and now that I think about it, should), but I hadn't gotten to this advanced of a use case yet. 😉

**Comment by isiahmeadows** [comment]
Okay, here's what my code looks like [without `RuleTester`](https://github.com/isiahmeadows/eslint-plugin-no-esnext/blob/c412813eb7301c0ade8254abe6ec515b84dfb4a9/test-util.js), so that's close to what I'd like to see with such an API added to core. (Mine is a little more specialized to my use case, so it obviously can't be directly ripped off.)

**Comment by platinumazure** [comment]
I would love to see something like: `ruleTester.runValid({ code, options, parserOptions })` and `ruleTester.runInvalid({ code, errors, options, parserOptions })` which would call `RuleTester.it` and handle one test case. Maybe also variants that don't call `RuleTester.it` and just run the testing logic/assertions.

This would allow some good options for these use cases and provide great code reuse, but without adding too much complexity or relying too heavily on conventions that might work in Mocha but maybe not in other test runners.

**Comment by isiahmeadows** [comment]
@platinumazure I'd find it easier if it didn't call `it` in the first place. (Alternatively, I could just set `RuleTester.it = (_, body) => body()` and it'd work just as well.)

**Comment by isiahmeadows** [comment]
The main boilerplate I'd like to elide is setting up the linter and then comparing the results appropriately. You'll still need a similar prototype, but it wouldn't try to run the rules in batch.
